### DISCUSSION TITLE
clarification on finalization

### DISCUSSION THREAD

**Discussion by euri10** [agronholm/anyio#753] [Q&A] [answered]
greetings, I first wrote this snippet, which appears to be an incorrect way of dealing with finalization:

```
@get("/foo")
async def foo(db_pool: Pool, db_connection: Connection) -> Stream:
    stream_id = await db_connection.fetchval(
        """insert into toto(stream) values ('') returning id"""
    )
    print(stream_id)

    async def chunks(num, stream_id):
        async with db_pool.acquire() as conn:
            try:
                for i in range(num):
                    chunk = f"chunk #{i}\n"
                    yield chunk
                    await asyncio.sleep(0.5)
                    update = await conn.execute(
                        f"update toto set stream = '{i}' where id={stream_id}"
                    )
                    print(update)
            except CancelledError as e:
                print(str(e))
                delete = await conn.execute(f"delete from toto where id={stream_id}")
                print(delete)
		raise
            finally:
                print("end")

    return Stream(chunks(10, stream_id))
```

then I read the docs which says:

> If you need to use await during finalization, you need to enclose it in a shielded cancel scope, or the operation will be cancelled immediately since it’s in an already cancelled scope:

so the "correct" way to write this would be to do:
```
            except CancelledError as e:
                print(str(e))
                with CancelScope(shield=True):
                    delete = await conn.execute(f"delete from toto where id={id}")
                    print(delete)
		raise
            finally:
                print("end")
```
and it indeed prints `DELETE 1` and effectively removes the entry fro the database.

however, it appears that the incorrect way of dealing with an await inside the exception deletes the database entry and doesnt print `DELETE 1`.

from the docs description I would have expected the delete NOT to be performed, so is there some part of the puzzle I'm missing here ?

**Comment by agronholm** [comment, marked as answer]
It all depends on what exactly `conn.execute()` does. It may or may not work, but gist is that if `CancelledError` was raised by an AnyIO cancel scope, then it will attempt to cancel the task over and over again until that `CancelledError` reaches the cancel scope that initiated the cancellation process.

**Reply by agronholm** [reply]
This is what `CancelScope(shield=True):` is for: it prevents that constant cancellation by AnyIO cancel scopes so that the code in `conn.execute()` can run without getting hit by those constant cancellations.

**Reply by euri10** [reply]
conn.execute is a simple asyncpg call, so if I understand your point it's working by accident in the case I didnt shield it, should the call to the db be "slower" then cancellation would happen before.

thanks for the clarification

**Comment by smurfix** [comment]
Could you please edit your question to fix the broken indent on the `raise` lines?

**Reply by agronholm** [reply]
> Could you please edit your question to fix the broken indent on the `raise` lines?

I did that now.

**Reply by euri10** [reply]
unless I'm missing something the docs put the raise at the same level than `with CancelScope(shield=True):`

**Reply by agronholm** [reply]
They were incorrectly indented in your paste, some 6 spaces deeper than the rest of the code.

**Reply by euri10** [reply]
respectfully, your edit doesnt match what the docs say or I'm completely braindead, which is possible today :rofl:

**Reply by agronholm** [reply]
WDYM? I just edited the code to make it valid Python syntax.

**Reply by euri10** [reply]
the docs put the raise at the same level than the with CancelledScope:
```
async def do_something():
    try:
        await run_async_stuff()
    except get_cancelled_exc_class():
        with CancelScope(shield=True):
            await some_cleanup_function()

        raise
```
you edit places it one tab before
```
            except CancelledError as e:
                print(str(e))
                with CancelScope(shield=True):
                    delete = await conn.execute(f"delete from toto where id={id}")
                    print(delete)
		raise
            finally:
                print("end")
```

**Reply by agronholm** [reply]
I don't know what you're seeing, but in the 2 snippets you just pasted, it's on the same level as the cancel scope.

**Reply by euri10** [reply]
would that be a github issue ?
printing a screen
![20240630_1712_1252x792_1719760341](https://github.com/agronholm/anyio/assets/1104190/daae94d5-c180-418c-8130-5ab7025b4ba8)

**Reply by agronholm** [reply]
This is what I'm seeing:
![image](https://github.com/agronholm/anyio/assets/130003/004f84c8-a0bd-4689-8b9c-812223758af7)

**Reply by agronholm** [reply]
I checked with a different browser too, and I'm seeing consistent results. And given that both myself and @smurfix saw a broken indent in your original paste leads me to believe that there is a problem with your browser that makes it display these snippets in an odd fashion.

**Reply by euri10** [reply]
ok this explains the misunderstanding, 
it's very worrying from github to say the least :)

**Reply by euri10** [reply]
just checked on firefox, which is the same for me as chrome though a little bit less indented, both on linux, could be because of a mix of tabs and spaces ?

anyways, that's not really the topic, but very confusing, both browsers are default one, no customization, stock debian bullseye, 

![20240630_1720_804x601_1719760840](https://github.com/agronholm/anyio/assets/1104190/04b66f68-a729-4396-9462-513d6cfa069a)

**Reply by agronholm** [reply]
I don't know what to say, except that I also checked on Windows (both Edge and Firefox) and the indent was rendered exactly the same there too.

**Reply by smurfix** [reply]
> could be because of a mix of tabs and spaces ?

Most probably. Lesson learned: Don't paste tabs.

**Reply by euri10** [reply]
ok sorry guys for this, i found the culprit in fact I'm an extension called Refined Github and removing it I see what I'm supposed to see :relieved: 

https://chromewebstore.google.com/detail/refined-github/hlepfoohegkhhmjieoechaddaejaokhf
